### ISSUE TITLE
2.6.12: Unable to resolve dependency: user requested 'io-console (= 0.4.5)'

### ISSUE THREAD

**Issue by matthewd** [rubygems/rubygems#1911]
After 2.6.12 was released, [our build started failing](https://travis-ci.org/rails/rails/jobs/227778564#L492) on a line that [ultimately](https://github.com/rails/rails/blob/00f1f95569a491c89bf78f08b78ebf0e6ff1d9d1/railties/lib/rails/generators/rails/app/templates/bin/setup.tt#L17) runs `gem install bundler --conservative`, and gets told:

```
ERROR:  While executing gem ... (Gem::UnsatisfiableDependencyError)
    Unable to resolve dependency: user requested 'io-console (= 0.4.5)'
```

My first guess is #1910, mostly because that seems the only pertinent change, but that's more "wild" than "well informed".

Looking at [the test](https://github.com/rails/rails/blob/00f1f95569a491c89bf78f08b78ebf0e6ff1d9d1/railties/test/application/bin_setup_test.rb#L40), I wonder if it would benefit from a `Bundler.with_clean_env`... but even if that would fix it, this change of behaviour seemed worth noting. It's also a pretty unclear error, IMO: none of our stuff mentions or touches `io-console`, so if we are in fact doing something newly-wrong (or always wrong but newly non-working), our mistake seems higher level than attempting to activate that particular gem.

Failed build (using 2.6.12): https://travis-ci.org/rails/rails/jobs/227778564
Last successful build (using 2.6.11): https://travis-ci.org/rails/rails/jobs/227369510

**Comment by segiddins** [comment]
Is there a series of commands I can use to reproduce this (outside of running rails)?

**Comment by matthewd** [comment]
I haven't worked out exactly what the test is doing that's special, but experimentation shows the following similar-looking (but maybe less confusingly bad) behaviour:

```
tango:src/rails[master]% bundle exec gem install bundler --conservative
ERROR:  While executing gem ... (Gem::UnsatisfiableDependencyError)
    Unable to resolve dependency: 'delayed_job (= 4.1.2)' requires 'activesupport (< 5.1, >= 3.0)'
```

I'm actually surprised bundler has seemingly allowed the lockfile to end up in its current state, and unlike the `io-console` case this involves comparatively relevant gems. But even granting those points, it seems surprising for a relatively low-level install operation to be complaining about satisfiability of something other than what we just asked to install (or one of its direct or indirect dependencies).

**Comment by segiddins** [comment]
And what's the `--backtrace`?

**Comment by segiddins** [comment]
So I've looked into this a bit, and it seems to result from using `bundler` with rails's gemfile in particular, since the Gemfile.lock seems to indicate we're using a version of activesupport which is indeed incompatible with the delayed job version...

**Comment by segiddins** [comment]
(My guess is the fix is going to be for bundler to make `Gem.finish_resolve` a no-op)

**Comment by segiddins** [comment]
```diff
diff --git a/lib/bundler/rubygems_integration.rb b/lib/bundler/rubygems_integration.rb
index 7283a326d..e33ae41b5 100644
--- a/lib/bundler/rubygems_integration.rb
+++ b/lib/bundler/rubygems_integration.rb
@@ -698,6 +698,10 @@ module Bundler
         Gem.post_reset do
           Gem::Specification.all = specs
         end
+
+        redefine_method((class << Gem; self; end), :finish_resolve) do |*|
+          []
+        end
       end
 
       def all_specs
diff --git a/spec/install/gemfile/gemspec_spec.rb b/spec/install/gemfile/gemspec_spec.rb
index e2534be1a..1ea613c9d 100644
--- a/spec/install/gemfile/gemspec_spec.rb
+++ b/spec/install/gemfile/gemspec_spec.rb
@@ -231,6 +231,30 @@ RSpec.describe "bundle install from an existing gemspec" do
     expect(the_bundle).to include_gems "foo 1.0.0"
   end
 
+  it "does not break Gem.finish_resolve with conflicts", :rubygems => ">= 2" do
+    build_lib("foo", :path => tmp.join("foo")) do |s|
+      s.version = "1.0.0"
+      s.add_dependency "bar", "= 1.0.0"
+    end
+    build_repo2 do
+      build_gem "deps" do |s|
+        s.add_dependency "foo", "= 0.0.1"
+      end
+      build_gem "foo", "0.0.1"
+    end
+
+    install_gemfile! <<-G
+      source "file://#{gem_repo2}"
+      gem "deps"
+      gemspec :path => '#{tmp.join("foo")}', :name => 'foo'
+    G
+
+    expect(the_bundle).to include_gems "foo 1.0.0"
+
+    run! "Gem.finish_resolve; puts 'WIN'"
+    expect(out).to eq("WIN")
+  end
+
   context "in deployment mode" do
     context "when the lockfile was not updated after a change to the gemspec's dependencies" do
       it "reports that installation failed" do

```

**Comment by segiddins** [comment]
This will be fixed with the next pre release of bundler 1.15
